src/actions/profile-view.js

```diff
@@ -149,6 +150,46 @@ export function selectLeafCallNode(
   };
 }
 
+/**
+ * Given a threadIndex and a sampleIndex, select the call node at the bottom ("root")
+ * of that sample's stack.
+ */
+export function selectRootCallNode(
+  threadIndex: ThreadIndex,
+  sampleIndex: IndexIntoSamplesTable
+): ThunkAction<void> {
+  return (dispatch, getState) => {
+    const threadSelectors = getThreadSelectors(threadIndex);
+    const filteredThread = threadSelectors.getFilteredThread(getState());
+    const callNodeInfo = threadSelectors.getCallNodeInfo(getState());
+
+    const newSelectedStack = filteredThread.samples.stack[sampleIndex];
+    const newSelectedCallNode =
+      newSelectedStack === null
+        ? -1
+        : callNodeInfo.stackIndexToCallNodeIndex[newSelectedStack];
+
+    const rootSelectedCallNode = findRootCallNode(
+      newSelectedCallNode,
+      callNodeInfo
+    );
+    dispatch(
+      changeSelectedCallNode(
+        threadIndex,
+        getCallNodePathFromIndex(
+          rootSelectedCallNode,
+          callNodeInfo.callNodeTable
+        ),
+        // Specify newSelectedCallNode as a descendant call node of rootSelectedCallNode.
+        getCallNodePathFromIndex(
+          newSelectedCallNode,
+          callNodeInfo.callNodeTable
+        )
+      )
+    );
```

Comment:
it may be easier to get the call node path from the index of the "new selected call node" first. A call node path is an array of function indexes. For example `[1, 2, 3]`. And in that case this is trivial to find the root: this is simply `[1]`.

So this is how this could be done:
```js
const selectedCallNodePath = getcallNodePathFromIndex(newSelectedCallNode, ...);
const rootCallNodePath = [selectedCallNodePath[0]];
dispatch(changeSelectedCallNode(threadIndex, rootCallNodePath, selectedCallNodePath);
```
I believe this is simpler :-) Tell me what you think!

This is one of the tricks in our code: there are 2 ways to point to call nodes, and the trick is to find the way that makes it easier!

Reply:
It is indeed simpler! Thanks for the explanation.